test/src/vfs_test_helper.h

```diff
@@ -0,0 +1,443 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file declares some vfs-specfic test suite helper functions.
+ */
+
+#ifndef TILEDB_VFS_TEST_HELPER_H
+#define TILEDB_VFS_TEST_HELPER_H
+
+#include "test/src/helpers.h"
+
+#ifdef _WIN32
+#include "tiledb/sm/filesystem/win.h"
+#else
+#include "tiledb/sm/filesystem/posix.h"
+#endif
+
+using namespace tiledb::common;
+
+namespace tiledb {
+
+namespace sm {}
+
+namespace test {
+
+/**
+ * This class defines and manipulates
+ * a list of supported filesystems.
+ */
+class SupportedFs {
+ public:
+  SupportedFs()
+      : temp_dir_()
+      , fs_vec() {
+  }
+
+  virtual ~SupportedFs() = default;
+  /* ********************************* */
+  /*               API                 */
+  /* ********************************* */
+
+  /**
+   * Returns Status upon setting up the associated
+   * filesystem's configuration
+   * Only for S3, Azure
+   * No-op otherwise
+   *
+   * @param config Configuration parameters
+   * @param error Error parameter
+   * @return Status OK if successful
+   */
+  virtual Status prepare_config(
+      tiledb_config_t* config, tiledb_error_t* error) = 0;
+
+  /**
+   * Creates bucket / container if does not exist
+   * Only for S3, Azure
+   * No-op otherwise
+   *
+   * @param ctx The TileDB context.
+   * @param vfs The VFS object.
+   * @return Status OK if successful
+   */
+  virtual Status init(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs) = 0;
+
+  /**
+   * Removes bucket / container if exists
+   * Only for S3, Azure
+   * No-op otherwise
+   *
+   * @param ctx The TileDB context.
+   * @param vfs The VFS object.
+   * @return Status OK if successful
+   */
+  virtual Status close(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs) = 0;
+
+  /**
+   * Get the name of the filesystem's directory
+   *
+   * @return string directory name
+   */
+  virtual std::string temp_dir() = 0;
+
+ private:
+  /* ********************************* */
+  /*           ATTRIBUTES              */
+  /* ********************************* */
+
+  /** The directory name of the associated filesystem. */
+  std::string temp_dir_;
```

Comment:
This does not appear to be used, we can remove it.

However: if we want to complicate things, we could make this `protected` instead of `private` and re-use it in the child classes. If we did this, we could provide the implementation for `virtual std::string temp_dir()` in the base class rather than re-implementing it for each child.

Reply:
@joe-maley which is preferred in this situation? I can do either

Reply:
I think it might be best to just remove this `temp_dir_` so that we can keep `SupportedFS` as a pure interface. While I was looking at this, I saw that the child classes all have their own `temp_dir_` alongside another `s3_temp_dir_` or whatever. The child classes just need one variable to store the temp dir, and then return it in their `temp_dir()` implementation.